### PR TITLE
Support tidbMonitor multiple replicas to ensure high availabilit

### LINKED ISSUE #2968: Support tidbMonitor multiple replicas to ensure high availability
## Feature Request

**Is your feature request related to a problem? Please describe:**
Should we support multiple replicas of TidbMonitor to ensure high availability?  Now ,the replica is one, hard-coded .

https://github.com/pingcap/tidb-operator/blob/e84a3d4fd007a62dfb3fd61681636357ebd787d3/pkg/monitor/monitor/util.go#L229
**Describe the feature you'd like:**

I think we can support replica field in tidbMonitor spec.

### PR DESCRIPTION
**Pull request by mikechengwei**
### What problem does this PR solve?
Close #2968 .

### What is changed and how does it work?
Support replicas field configuration in tidbmonitor.

### Code changes

- [ ] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [ ] E2E test 
- [ ] Manual test 
### Release Notes

```release-note
```

### CODE REVIEW

File: pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go [hunk 1]
```diff
@@ -105,6 +105,12 @@ type TidbMonitorSpec struct {
 	// Defaults to the value of `prometheus_replica`. External label will
 	// _not_ be added when value is set to empty string (`""`).
 	ReplicaExternalLabelName *string `json:"replicaExternalLabelName,omitempty"`
+	// Replicas is the number of desired replicas.
+	// This is a pointer to distinguish between explicit zero and unspecified.
+	// Defaults to 1.
+	// More info: https://kubernetes.io/docs/concepts/workloads/controllers/replicationcontroller#what-is-a-replicationcontroller
+	// +optional
+	Replicas *int32 `json:"replicas,omitempty" protobuf:"varint,1,opt,name=replicas"`
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go:113]:
```suggestion
	// Replicas is the number of desired replicas.
	// Defaults to 1.
	// +optional
	Replicas *int32 `json:"replicas,omitempty"`
```

**Comment by DanielZhangQD** [comment]
@mikechengwei Do the following services work normally with multiple replicas?

- Grafana
- Prometheus
- Reloader

**Comment by mikechengwei** [comment]
> @mikechengwei Do the following services work normally with multiple replicas?
> 
> * Grafana
> * Prometheus
> * Reloader

Of course. The e2e test is ok.

**Comment by DanielZhangQD** [comment]
> > @mikechengwei Do the following services work normally with multiple replicas?
> > 
> > * Grafana
> > * Prometheus
> > * Reloader
> 
> Of course. The e2e test is ok.

@mikechengwei The reloader is used to save the customized alert rules for Prometheus, and the customized rules will be saved in the PVCs, if with two Pods, when we customize the rules with reloader service, the rules will be saved only in one Pod/PVC, the rules in the other one will be untouched, how to solve this?
BTW, with two Prometheus Pods, two duplicate alerts will be fired for one alert event?

**Comment by mikechengwei** [comment]
> > > @mikechengwei Do the following services work normally with multiple replicas?
> > > 
> > > * Grafana
> > > * Prometheus
> > > * Reloader
> > 
> > 
> > Of course. The e2e test is ok.
> 
> @mikechengwei The reloader is used to save the customized alert rules for Prometheus, and the customized rules will be saved in the PVCs, if with two Pods, when we customize the rules with reloader service, the rules will be saved only in one Pod/PVC, the rules in the other one will be untouched, how to solve this?
> BTW, with two Prometheus Pods, two duplicate alerts will be fired for one alert event?

First of all, you have to make it clear that this is for high availability of monitoring. As for configuration consistency, we should solve it from the operator's perspective. Alertmanager is to configure a prometheus service ip.

**Comment by DanielZhangQD** [comment]
Currently, the alert manager URL is configured in TidbMonitor:
```
spec:
  alertmanagerURL: http://alertmanager-main.monitoring:9093
```
And it will finally be populated to the Prometheus config:
```
    alerting:
      alertmanagers:
      - static_configs:
        - targets:
          - http://alertmanager-main.monitoring:9093
```
In this case, two alerts will be reported for one alert, right?

**Comment by mikechengwei** [comment]
> Currently, the alert manager URL is configured in TidbMonitor:
> 
> ```
> spec:
>   alertmanagerURL: alertmanager-main.monitoring:9093
> ```
> 
> And it will finally be populated to the Prometheus config:
> 
> ```
>     alerting:
>       alertmanagers:
>       - static_configs:
>         - targets:
>           - alertmanager-main.monitoring:9093
> ```
> 
> In this case, two alerts will be reported for one alert, right?

You're right. But if we adopt the thanos solution, there is no such concern,thanos ruler is running alone.
And you can see replica is support in prometheus operator.
https://github.com/prometheus-operator/prometheus-operator/blob/322580894cb0eb2cacf26ae9defcacd23abc8df2/pkg/apis/monitoring/v1/types.go#L138

**Review by DanielZhangQD — APPROVED** [review]
LGTM

**Pull request merged**